crates/ruff_python_parser/src/lexer.rs

```diff
@@ -552,7 +552,10 @@ impl<'source> Lexer<'source> {
 
         loop {
             match self.cursor.first() {
-                EOF_CHAR => {
+                // The condition is to differentiate between the `NUL` (`\0`) character
+                // in the source code and the one returned by `self.cursor.first()` when
+                // we reach the end of the source code.
+                EOF_CHAR if self.cursor.is_eof() => {
```

Comment:
We could move this in the catch-all branch but that would mean that the check will be performed for almost _every_ character.

Reply:
Yeah, that's the only place where it seems necessary.